File: codes/07/VMTranslator/CodeWriter.java
import java.io.BufferedWriter;
import java.io.FileWriter;
import java.io.IOException;

public class CodeWriter {
    private int jumpCounter = 0;
    private BufferedWriter outPrinter;

    public CodeWriter(String writer) throws IOException {
        this.outPrinter = new BufferedWriter(new FileWriter(writer));
    }

    public void close() throws IOException {
        outPrinter.close();
    }

    public void writeAssembly(CmdType cmdType, String arg1, Integer arg2) throws IOException {
        String asmCode = "";
        switch (cmdType) {
            case CArithmetic:
                asmCode = writeArithmetic(arg1);
                break;
            case CPush:
                asmCode = writePush(arg1, arg2);
                break;
            case CPop:
                asmCode = writePop(arg1, arg2);
                break;
            default:
                throw new UnsupportedOperationException("Command type not supported.");
        }
        outPrinter.write(asmCode);
    }

    private String getSegment(String segment, int index) {
        switch (segment) {
            case "local":
                return "LCL";
            case "argument":
                return "ARG";
            case "this":
                return "THIS";
            case "that":
                return "THAT";
            case "pointer":
                return (index == 0) ? "THIS" : "THAT";
            case "static":
                return Integer.toString(16 + index);
            default:
                throw new IllegalArgumentException("Invalid segment");
        }
    }

    private boolean getDirect(String segment) {
        return segment.equals("pointer") || segment.equals("static");
    }

    private String writeArithmetic(String command) {
        String result = "";
        switch (command) {
            case "add":
                result = arithmeticTemplate1() + "M=M+D\n";
                break;
            case "sub":
                result = arithmeticTemplate1() + "M=M-D\n";
                break;
            case "and":
                result = arithmeticTemplate1() + "M=M&D\n";
                break;
            case "or":
                result = arithmeticTemplate1() + "M=M|D\n";
                break;
            case "gt":
            case "lt":
            case "eq":
                String jumpType = command.equals("gt") ? "JLE" :
                                  command.equals("lt") ? "JGE" : "JNE";
                jumpCounter++;
                result = arithmeticTemplate2(jumpType);
                break;
            case "not":
                result = "@SP\nA=M-1\nM=!M\n";
                break;
            case "neg":
                result = "D=0\n@SP\nA=M-1\nM=D-M\n";
                break;
            default:
                throw new UnsupportedOperationException("Arithmetic command not supported.");
        }
        return result;
    }

    private String writePush(String segment, int index) {
        if (segment.equals("constant")) {
            return String.format("@%d\nD=A\n@SP\nA=M\nM=D\n@SP\nM=M+1\n", index);
        }
        String seg;
        boolean isDirect;
        if (segment.equals("temp")) {
            seg = "R" + (5 + index);
            isDirect = true;
        } else {
            seg = getSegment(segment, index);
            isDirect = getDirect(segment);
        }
        return pushTemplate(seg, index, isDirect);
    }

    private String writePop(String segment, int index) {
        if (segment.equals("temp")) {
            return String.format("@SP\nAM=M-1\nD=M\n@R%d\nM=D\n", 5 + index);
        }
        String seg;
        boolean isDirect;
        if (segment.equals("static")) {
            seg = String.valueOf(16 + index);
            isDirect = true;
        } else {
            seg = getSegment(segment, index);
            isDirect = getDirect(segment);
        }
        return popTemplate(seg, index, isDirect);
    }

    private String arithmeticTemplate1() {
        return "@SP\nAM=M-1\nD=M\nA=A-1\n";
    }

    private String arithmeticTemplate2(String jumpType) {
        return String.format("@SP\nAM=M-1\nD=M\nA=A-1\nD=M-D\n@FALSE%d\nD;%s\n@SP\nA=M-1\nM=-1\n@CONTINUE%d\n0;JMP\n(FALSE%d)\n@SP\nA=M-1\nM=0\n(CONTINUE%d)\n",
                jumpCounter, jumpType, jumpCounter, jumpCounter, jumpCounter);
    }

    private String pushTemplate(String segment, int index, boolean isDirect) {
        String indexCode = isDirect ? "" : String.format("@%d\nA=D+A\nD=M\n", index);
        return String.format("@%s\nD=M\n%s@SP\nA=M\nM=D\n@SP\nM=M+1\n", segment, indexCode);
    }

    private String popTemplate(String segment, int index, boolean isDirect) {
        String pointerAdjust = isDirect ? "D=A\n" : String.format("D=M\n@%d\nD=D+A\n", index);
        return String.format("@%s\n%s@R13\nM=D\n@SP\nAM=M-1\nD=M\n@R13\nA=M\nM=D\n", segment, pointerAdjust);
    }
}

File: codes/07/vmtranslator/src/code_writer.rs
use std::fs::File;
use std::io::{self, BufWriter, Write};
use crate::cmd_type::CmdType;

pub(crate) struct CodeWriter {
    jump_counter: i32,
    out_printer: BufWriter<File>,
}

impl CodeWriter {
    pub fn new(file_out: &str) -> io::Result<Self> {
        let file = File::create(file_out)?;
        let out_printer = BufWriter::new(file);
        Ok(CodeWriter {
            jump_counter: 0,
            out_printer,
        })
    }

    pub fn write_assembly(&mut self, cmd_type: CmdType, arg1: Option<String>, arg2: Option<i32>) -> io::Result<()> {
        let asm_code = match cmd_type {
            CmdType::CArithmetic => self.write_arithmetic(&arg1.unwrap()),
            CmdType::CPush => self.write_push(&arg1.unwrap(), arg2.unwrap()),
            CmdType::CPop => self.write_pop(&arg1.unwrap(), arg2.unwrap()),
            _ => unreachable!(),
        };
        self.out_printer.write_all(asm_code.as_bytes())?;
        Ok(())
    }

    fn get_segment_and_direct(&self, segment: &str, index: i32) -> (String, bool) {
        match segment {
            "local" => ("LCL".to_string(), false),
            "argument" => ("ARG".to_string(), false),
            "this" => ("THIS".to_string(), false),
            "that" => ("THAT".to_string(), false),
            "pointer" => {
                let seg = if index == 0 { "THIS" } else { "THAT" };
                (seg.to_string(), true)
            }
            "static" => (format!("{}", 16 + index), true),
            _ => panic!("Invalid segment"),
        }
    }

    fn write_arithmetic(&mut self, command: &str) -> String {
        match command {
            "add" => format!("{}M=M+D\n", CodeWriter::arithmetic_template1()),
            "sub" => format!("{}M=M-D\n", CodeWriter::arithmetic_template1()),
            "and" => format!("{}M=M&D\n", CodeWriter::arithmetic_template1()),
            "or" => format!("{}M=M|D\n", CodeWriter::arithmetic_template1()),
            "gt" | "lt" | "eq" => {
                let jump_type = match command {
                    "gt" => "JLE",
                    "lt" => "JGE",
                    "eq" => "JNE",
                    _ => unreachable!(),
                };
                self.jump_counter += 1;
                format!("{}", self.arithmetic_template2(jump_type))
            },
            "not" => "@SP\nA=M-1\nM=!M\n".to_string(),
            "neg" => "D=0\n@SP\nA=M-1\nM=D-M\n".to_string(),
            _ => unreachable!(),
        }
    }

    fn write_push(&self, segment: &str, value: i32) -> String {
        match segment {
            "constant" => format!("@{}\nD=A\n@SP\nA=M\nM=D\n@SP\nM=M+1\n", value),
            "temp" => format!("@R{}\nD=M\n@SP\nA=M\nM=D\n@SP\nM=M+1\n",value + 5),
            _ => {
                let (seg, is_direct) = self.get_segment_and_direct(segment, value);
                self.push_template(&seg, value, is_direct)
            }
        }
    }

    fn write_pop(&self, segment: &str, value: i32) -> String {
        match segment {
            "temp" => format!("@SP\nAM=M-1\nD=M\n@R{}\nM=D\n", value + 5),
            _ => {
                let (seg, is_direct) = self.get_segment_and_direct(segment, value);
                self.pop_template(&seg, value, is_direct)
            }
        }
    }

    fn arithmetic_template1() -> String {
        "@SP\nAM=M-1\nD=M\nA=A-1\n".to_string()
    }

    fn arithmetic_template2(&self, jump_type: &str) -> String {
        format!(
            "@SP\n\
             AM=M-1\n\
             D=M\n\
             A=A-1\n\
             D=M-D\n\
             @FALSE{}\n\
             D;{}\n\
             @SP\n\
             A=M-1\n\
             M=-1\n\
             @CONTINUE{}\n\
             0;JMP\n\
             (FALSE{})\n\
             @SP\n\
             A=M-1\n\
             M=0\n\
             (CONTINUE{})\n",
            self.jump_counter, jump_type, self.jump_counter, self.jump_counter, self.jump_counter
        )
    }

    fn push_template(&self, segment: &str, index: i32, is_direct: bool) -> String {
        let no_pointer_code = if is_direct {
            "".to_string()
        } else {
            format!("@{}\nA=D+A\nD=M\n", index)
        };
        format!(
            "@{}\n\
             D=M\n\
             {}\
             @SP\n\
             A=M\n\
             M=D\n\
             @SP\n\
             M=M+1\n",
            segment, no_pointer_code
        )
    }

    fn pop_template(&self, segment: &str, index: i32, is_direct: bool) -> String {
        let no_pointer_code = if is_direct {
            "D=A\n".to_string()
        } else {
            format!("D=M\n@{}\nD=D+A\n", index)
        };
        format!(
            "@{}\n\
             {}\
             @R13\n\
             M=D\n\
             @SP\n\
             AM=M-1\n\
             D=M\n\
             @R13\n\
             A=M\n\
             M=D\n",
            segment, no_pointer_code
        )
    }
}
